semgrep-core/src/analyzing/Constant_propagation.ml

```diff
@@ -563,17 +562,25 @@ let propagate_basic lang prog =
                 attrs;
                 _;
               },
-              VarDef { vinit = Some { e = L literal; _ }; _ } )
+              VarDef { vinit = Some e; _ } )
           (* note that some languages such as Python do not have VarDef.
            * todo? should add those somewhere instead of in_lvalue detection?*)
             -> (
               match Hashtbl.find_opt stats (H.str_of_ident id, sid) with
               | Some stats ->
-                  if
-                    H.has_keyword_attr Const attrs
-                    || H.has_keyword_attr Final attrs
-                    || (!(stats.lvalue) = 1 && is_js env)
-                  then add_constant_env id (sid, Lit literal) env;
+                  (if
+                   H.has_keyword_attr Const attrs
+                   || H.has_keyword_attr Final attrs
+                   || !(stats.lvalue) = 1
+                  then
+                   match e.e with
+                   | L literal -> add_constant_env id (sid, Lit literal) env
+                   (* For any other expression, it is OK to propagate it symbolically so long as
+                      the lvalue is only assigned to once.
+                      Although we may propagate expressions with identifiers in them, those identifiers
+                      will simply not have an `svalue` if they are non-propagated as well.
+                   *)
+                   | _ -> add_constant_env id (sid, Sym e) env);
```

Comment:
At least for now I prefer that we restrict this to `Dataflow_svalue.is_symbolic_expr` same way as the dataflow analysis does.